abnf/odata-aggregation-abnf.txt

```diff
@@ -130,15 +130,25 @@ pathPrefix      = [ ( qualifiedEntityTypeName / qualifiedComplexTypeName ) "/" ]
 pathSegment     = ( complexProperty / complexColProperty ) [ "/" qualifiedComplexTypeName ]
                 / navigationProperty [ "/" qualifiedEntityTypeName  ]
 
-groupingProperty = grpPropPathPrefix
-                   ( entityNavigationProperty [ "/" qualifiedEntityTypeName ]
-                   / primitiveProperty 
-                   / complexProperty
-                   )
-grpPropPathPrefix  = [ ( qualifiedEntityTypeName / qualifiedComplexTypeName ) "/" ] *( grpPropPathSegment "/" )
-grpPropPathSegment = complexProperty          [ "/" qualifiedComplexTypeName ]
-                   / entityNavigationProperty [ "/" qualifiedEntityTypeName  ]
-                   
+; a type cast (qualifiedComplexTypeName or qualifiedEntityTypeName) cannot be the final segment
+groupingProperty   = [ ( qualifiedComplexTypeName / qualifiedEntityTypeName ) "/" ]
+                     ( grpPrimPathSegment / grpPropPathSegment )
+grpPropPathSegment = grpPropAnnoSegment
+                   / complexProperty          [ [ "/" qualifiedComplexTypeName ] "/" grpPropPathSegment ]
+                   / entityNavigationProperty [ [ "/" qualifiedEntityTypeName  ] "/" grpPropPathSegment ]
+                   / primitiveProperty                                           "/" grpPropAnnoSegment
+grpPrimPathSegment = grpPrimAnnoSegment
+                   / complexProperty          [ [ "/" qualifiedComplexTypeName ] "/" grpPrimPathSegment ]
+                   / entityNavigationProperty [ [ "/" qualifiedEntityTypeName  ] "/" grpPrimPathSegment ]
+                   / grpPropPathSegment                                          "/" grpPrimPathSegment
+                   / primitiveProperty                                           "/" grpPrimAnnoSegment
+                   / primitiveProperty
+grpPropAnnoSegment = complexAnnotationInQuery [ [ "/" qualifiedComplexTypeName ] "/" grpPropPathSegment ]
+                   / entityAnnotationInQuery  [ [ "/" qualifiedEntityTypeName  ] "/" grpPropPathSegment ]

```

Comment:
Must this really be so much longer than the somewhat similar `expandPath`)?
https://github.com/oasis-tcs/odata-abnf/blob/1c775eb1e6609a7bc1d3502283634cdb63854c02/abnf/odata-abnf-construction-rules.txt#L285-L293